iml-agent/src/action_plugins/stratagem/action_warning.rs

```diff
@@ -81,3 +88,50 @@ pub fn write_records(
 
     Ok(())
 }
+
+pub fn read_mailbox(
+    device: &str,
+    mailbox: &str,
+    out: impl io::Write,
+) -> Result<(), ImlAgentError> {
+    let mntpt = liblustreapi::search_rootpath(&device).map_err(|e| {
+        log::error!("Failed to find rootpath({}) -> {:?}", device, e);
+        e
+    })?;
+
+    let query: Vec<(String, String)> = Vec::new();
+    let mut wtr = csv::Writer::from_writer(out);
+    let message_endpoint = env::MANAGER_URL.join("/mailbox/")?.join(mailbox)?;
+    let identity = crypto_client::get_id(&env::PFX)?;
+    let client = crypto_client::create_client(identity)?;
+
+    let (sender, recv) = channel();
+
+    // Spawn off an expensive computation
+    tokio::spawn(
+        stream_lines::strings(crypto_client::get_stream(&client, message_endpoint, &query))
+            .for_each(move |fid| {
+                if let Ok(rec) = fid2record(&mntpt, &fid) {
```

Comment:
The `for_each` call here will run `fid2record` in a serial way.

I think this is where you may benefit from `rayon-futures`.

Reply:
Or perhaps rayon is overkill and you can just wrap this call in another `tokio::spawn` (to create a new parallel task), with the `blocking` annotation. I think that may end up self-limiting once blocking threadpool is used up.

In either case, it's probably worth throwing a large dataset at this to see how it performs.

Reply:
Example of blocking call:

https://github.com/whamcloud/integrated-manager-for-lustre/blob/a0067c7d4debdec9782bec57b6164cfe43561a1f/iml-agent/src/fs.rs#L80-L94

Reply:
Does the `use rayon::prelude::*;` at the top not cause the `for_each` to be parallel?

Reply:
I don't think so, according to https://dev.azure.com/whamcloud/Integrated%20Manager%20For%20Lustre/_build/results?buildId=683 the import is unused:

```
warning: unused import: `rayon::prelude`
  --> iml-agent/src/action_plugins/stratagem/action_warning.rs:16:5
   |
16 | use rayon::prelude::*;
   |     ^^^^^^^^^^^^^^
```

Reply:
I guess it's "automatic" if you use par_iter() in place of an iter() call.